### ISSUE TITLE
safely destroying semaphore used with WSI

### ISSUE THREAD

**Issue by jjulianoatnv** [KhronosGroup/Vulkan-Docs#152]
This issue seeks clarification on when it is safe to destroy a semaphore that has been used by WSI functions vkAcquireNextImageKHR and vkQueuePresentKHR.

A separate issue in the layer repo will address the validation layer detecting when a semaphore used by these functions is destroyed too soon. (I plan to edit this text with a cross-reference after the layer issue has been filed.)

---

vkAcquireNextImage's semaphore is used to establish or extend a dependency chain with a write-after-read hazard ensuring that the acquired image is modified only after the presentation engine is done reading from the image.

Quoting from the spec: 

> If semaphore is not VK_NULL_HANDLE, the semaphore must be unsignaled and not have any uncompleted signal or wait operations pending. It will become signaled when the presentation engine has released ownership of the image, and the device may modify its contents. Queue operations that access the image contents must wait until the semaphore signals; typically applications should include the semaphore in the pWaitSemaphores list for the queue submission that transitions the image away from the VK_IMAGE_LAYOUT_PRESENT_SRC_KHR layout. Use of the semaphore allows rendering operations to be recorded and submitted before the presentation engine releases ownership.

and

> A successful call to vkAcquireNextImageKHR counts as a signal operation on semaphore for the purposes of queue forward-progress requirements. The semaphore is guaranteed to signal, so a wait operation can be queued for the semaphore without risk of deadlock.

This seems to fully defines when vkAcquireNextImageKHR's semaphore may be destroyed. I think it says that semaphore may be destroyed after completion of all queue operations that include semaphore in the pWaitSemaphores list for the queue submission. But that destroying it any sooner results in undefined behavior. In other words, one must track completion of all consumers of semaphore to know when it is safe to destroy semaphore.

For the validation layer to detect an application destroying vkAcquireNextImageKHR's semaphore too soon, I think the burden is no different than for other semaphores that are not used by vkAcquireNextImageKHR.

---

Switching to vkQueuePresentKHR, pPresentInfo->pWaitSemaphores extends a dependency chain with a read-after-write hazard ensuring that the images are presented only after writes to the images have completed.

Quoting from the spec:

> pWaitSemaphores, if non-NULL, is an array of VkSemaphore objects with waitSemaphoreCount entries, and specifies the semaphores to wait for before issuing the present request.

and

> The transfer of ownership to the presentation engine happens in issue order with other queue operations, but semaphores have to be used to ensure that prior rendering and other commands in the specified queue complete before the presentation operation. The presentation operation itself does not delay processing of subsequent commands on the queue, however, presentation requests sent to a particular queue are always performed in order. Exact presentation timing is controled by the semantics of the presentation engine and native platform in use.

Comparing these two excerpts there might be an issue here with the specific spec language of "present request" and "present operation". I'm addressing that separately within Khronos. I will assume that vkQueuePresentKHR is the present request, and that the present operation is carried out by the presentation engine after the present request schedules on the present operation on the presentation engine's own queue. (The presentation engine's queue is separate from the VkQueue.)

The spec excerpts do not seem to fully define when the semaphores pointed to by pWaitSemaphores may be destroyed. It is my interpretation of both the spec language and the intent of the spec that a non-blocking vkQueuePresentKHR request may schedule a present operation on the presentation engine's queue such that the present operation on the presentation engine's queue has a read-after-write hazard on pWaitSemaphores. This would mean that pWaitSemaphores may not be destroyed until after the presentation engine has completed the present operation.

However there seems to be nothing in vkQueuePresentKHR to indicate to the app when the presentation engine has completed the present operation. The next-best thing is to assume that pWaitSemaphires may be destroyed after the swapchain is destroyed, or after a future vkAcquireNextImageKHR returns the same image. The latter imposes a book-keeping requirement on the application that is non-obvious and that had not occurred to me before now.

It might be possible to assume the semaphores may be destroyed after vkAcquireNextImageKHR returns some number of other images. How many? I don't see language in the spec in support of such an assumption. I'm not even entirely certain that the spec supports the earlier assumption of "the same image".

---

Finally, the purpose of this issue is to feel out and document discussion related to clarifying the spec language defining when the application may destroy the semaphores pointed to by vkQueuePresentKHR's pWaitSemaphores. If the existing spec language already covers this adequately that's fine, and this issue can document that interpretation.

**Comment by jjulianoatnv** [comment]
However this is resolved, support in the validation layers is needed. One of the factors motivating creation of this issue is that I don't see how to add this to the validation layer without first clarifying when it is safe to destroy the semaphores pointed to by vkQueuePresentKHR's pWaitSemaphores.

**Comment by timofonic** [comment]
@jjulianoatnv @oddhack Any news?

**Comment by krOoze** [comment]
I think it is mostly there since the synchronization descriptions overhaul. Some final touches in #785.

**Comment by krOoze** [comment]
Yea, I do somewhat keep finding loopholes and ambiguities.

What worries me most is this spec quote:

> Also, in the case of `vkQueueSubmit`, the second synchronization scope additionally includes all commands that occur later in submission order.

(Same for Semaphore and Fence.) It marks any other queue commands that are not `vkQueueSubmit` to not include what follows in submission order. It somehow makes sense that we do not want to stall rest of the work, but it makes things much harder.

I had some time to collect my thoughts making the PR, and I think the (hopefully complete) behavior loosely based my current spec interpretation is (while necesserily making some stuff up):
- acquire with a Fence is hopefully straigtforward. If you wait on the fence, then if there is also a semaphore, then this does not prevent destroying the semaphore, destroying the swapchain, and the image is already _visible_ to whatever you want on device domain.
- acquire with only Semaphore is done when you wait on the Semaphore, and know the waiting stuff has finished execution (by regular means). Or if you call a `vkDeviceWaitIdle`. `vkQueueWaitIdle` should **not** work no matter the used Queue (or even if used on all Queues known to the user via API).
  - it does **not** cover any work earlier in submission order. Except in the case the acquired image is an previously presented image. In that case the batch of the present is considered finished (and semaphore safe to destroy, and stuff that signalled this semaphore finished execution)
- present with a semaphore does not cover subsequent submissions. We know the work is done (and semaphore destroyable etc) if we call `vkDeviceWaitIdle`, `vkQueueWaitIdle`, different semaphore from subsequent `vkQueueSubmit` (which covers previous queue submissions) was signaled, or if we know Present following this present has finished execution.

Does that make sense? I welcome any notes. It also feels awfully complex. I wonder if I am overthinking it, and there is a room to make this much simpler.

**Comment by cubanismo** [comment]
> acquire with only Semaphore is finished when you wait on the Semaphore, and know the waiting stuff has finished execution (by regular means). **Or if you call a vkDeviceWaitIdle.** vkQueueWaitIdle should not work no matter the used Queue (or even if used on all Queues known to the user via API).

In the above (emphasis mine) statement, I don't think the bold portion is true.  DeviceWaitIdle does not cover the pending work that signals the semaphore, because that work happens external to Vulkan.  It is presentation engine work.  Hence, the only way to ensure the semaphore is idle is to first wait on it.  After that, the usual synchronization logic falls out naturally because at that point, you have work related to the semaphore enqueued on a well-defined device queue.

From my understanding, the rest of the assertions in the above comment are correct.

**Comment by cubanismo** [comment]
Think about how WSI would be/is implemented.  It's essentially an under-the-covers external object situation.  If you had a semaphore shared between two VkDevice objects, potentially in separate processes, would you expect calling vkDeviceWaitIdle() on one of those devices to idle any pending operations on the semaphore happening on the other device?  That would be an impossible implementation requirement.

The caveat with this line of thinking is that in such a case, there would be two OS-level references to the semaphore payload, so you could safely destroy one instance of the semaphore without idling the other.  This gets fuzzy with AcquireNextImage.  It essentially binds the specified semaphore to an opaquely-defined semaphore payload whose lifetime is not discussed in the spec as far as I can tell.  Without examining our driver more closely, I'm not even sure what I can say about our implementation's behavior if the device were destroyed while operations on such semaphores are pending.  This ambiguity is unfortunate, but it doesn't change the awkwardness of asking the driver to idle things happening outside of the scope of its queues.

**Comment by krOoze** [comment]
@cubanismo No, I would expect only  the device that user asked to be stopped to be stopped. `VkSwapchainKHR` is a child of `VkDevice`, therefore I (ideally) expect it to be stopped.

I mean, you are talking about implementation details. You could say the same about `vkDestroySemaphore`. How do I know I am not destroying semaphore that some other process is using? Simple: it is not my resposibility. If the Implementation uses it that way, the Implementation has to clean it in whatever way is appropriate.

Anyway, I am gonna make an Issue as it seems to be a separate can of worms that actually might impact compatibility.

**Comment by krOoze** [comment]
BTW, the part I am most worried is actually:

> different semaphore from subsequent vkQueueSubmit (which covers previous queue submissions) was signaled, or if we know Present following this present has finished execution on the same queue.

Because `vkQueuePresent` is not formally part of the submission order.

**Comment by krOoze** [comment]
I think Signal Operation Order tries to say something here. With some creative interpretation I think it makes valid:

```
VkQueue q;
VkSemaphore s;
vkQueuePresent( q, s );
VkFence f;
vkQueueSubmit( q, f );
vkWaitForFences( f );
vkDestroySemaphore( s );
```

@oddhack Can whoever wrote Signal Operation Order in 1.1.124 weigh in here?

**Comment by krOoze** [comment]
Any update on this?
As this is significant gap in Vulkan system, may I ask this Issue to be prioritized.

So far only decent and fail-proof answer I got is the semaphore is considered retired when subsequent `vkAcquireNextImage` returns the same swapchain image and you wait on its fence. But that is not viable strategy for the purpose of semaphore destruction, as you cannot keep calling `vkAcquireNextImage`s for no reason.

**Comment by krOoze** [comment]
It is still pretty large hole in core Vulkan, I think. Was this discussed at Khronos? cc @oddhack

**Comment by oddhack** [comment]
> It is still pretty large hole in core Vulkan, I think. When can one destroy\reuse semaphores passed to WSI?? Was this discussed at Khronos? cc @oddhack

@cubanismo do we have an internal tracker for this? I haven't been in the WSI discussions to know if it's come up or not.

**Comment by krOoze** [comment]
@cubanismo poke

**Comment by cubanismo** [comment]
There is work pending within Khronos to address this issue via a new extension. I don't have any specific updates on that work at this time.

**Comment by krOoze** [comment]
>  to address this issue via a new extension

Do you also plan to specify how to deal with this **without** using any new extensions?

**Comment by HansKristian-Work** [comment]
For the QueuePresentKHR, the spec has wording:

> Queueing an image for presentation defines a set of queue operations, including waiting on the semaphores and submitting a presentation request to the presentation engine. However, the scope of this set of queue operations does not include the actual processing of the image by the presentation engine.

So this means that a QueueWaitIdle, or submitting a fence -> wait ensures that the semaphore(s) have been waited on. The part which happens outside the queue is the actual "pixel hitting screen" thing.

Even for typical WSI errors, this holds:

> However, if the presentation request is rejected by the presentation engine with an error VK_ERROR_OUT_OF_DATE_KHR, VK_ERROR_FULL_SCREEN_EXCLUSIVE_MODE_LOST_EXT, or VK_ERROR_SURFACE_LOST_KHR, the set of queue operations are still considered to be enqueued and thus any semaphore wait operation specified in VkPresentInfoKHR will execute when the corresponding queue operation is complete.

AcquireNextImageKHR is the awkward one indeed. I think the only way in current spec is to submit a wait on the semaphore in some queue, then wait idle on that.

**Comment by krOoze** [comment]
> So this means that a QueueWaitIdle, or submitting a fence -> wait ensures that the semaphore(s) have been waited on. The part which happens outside the queue is the actual "pixel hitting screen" thing.

It means nothing, because the source scope of a fence signal operation does not include presentation. So even after `vkQueueWaitIdle` the present semaphore state is indeterminate afterwards.

`vkAcquireNextImageKHR` AIS above is highly inconvenient, but seems properly specified as semaphore import operation. An explicit warning note should perhaps be added though that it has to be waited on to sanitize it.

**Comment by HansKristian-Work** [comment]
> does not include presentation request batch, nor the semaphore unsignal op

A fence signal operation waits for all queue operations, so I don't see why semaphore unsignal wouldn't be part of that.

**Comment by krOoze** [comment]
It does not wait for "all queue operations" as I have quoted out above. I have to assume it waits for what the spec says it waits for.

**Comment by HansKristian-Work** [comment]
I'm reading this:
> Before a fence or semaphore is signaled, it is guaranteed that any previously submitted queue operations have completed execution

And that would include wording for QueuePresentKHR which states that waiting on the semaphore is a queue operation.

**Comment by krOoze** [comment]
That is introductory\informative chapter. Later on you have formal specification of [fence signal op](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#synchronization-fences-signaling).

**Comment by HansKristian-Work** [comment]
> Fence signal operations that are defined by vkQueueSubmit or vkQueueBindSparse additionally include in the first synchronization scope any semaphore and fence signal operations that occur earlier in signal operation order.

Does vkQueuePresentKHR somehow escape "any" here?

**Comment by krOoze** [comment]
Yes, `vkQueuePresentKHR` is not a fence signal op.

**Comment by HansKristian-Work** [comment]
> Yes, `vkQueuePresentKHR` is not a fence signal op.

I read it as any semaphore operation, and fence signal operation.

**Comment by krOoze** [comment]
Doesn't matter how you read it. Only thing that matters is what is written.

Either way, `vkQueuePresentKHR` isn't part of signal order neither.

**Comment by HansKristian-Work** [comment]
So reading through https://github.com/KhronosGroup/Vulkan-Docs/issues/1150, and somehow PresentKHR was explicitly removed from signal order indeed. Given how Submit/BindSparse live in somewhat different sync domains, I think I buy that Present also lives separately, so a VkFence is out ... I really shouldn't do sync stuff at 3 am ...

I don't think we have ruled out QueueWaitIdle working though.

> To wait on the host for the **completion of outstanding queue operations** for a given queue, call:

The following wording is unfortunate though:
> vkQueueWaitIdle is equivalent to submitting a fence to a queue and waiting with an infinite timeout for that fence to signal.

That seems not well specified even for just sparse.

**Comment by krOoze** [comment]
We have not ruled in `vkQueueWaitIdle` working neither. At best the spec is ambiguous and things it says are incompatible with each other. Hence this Issue.

**Comment by devshgraphicsprogramming** [comment]
I have a more important question, do I need do present all the images I've acquired before I destroy a swapchain?

**Comment by ShabbyX** [comment]
The new [VK_EXT_surface_maintenance1][surface] and [VK_EXT_swapchain_maintenance1][swapchain] extensions address this issue. Please see this [blog post][blog] for a summary.

[surface]: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_surface_maintenance1.html
[swapchain]: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_swapchain_maintenance1.html
[blog]: https://www.khronos.org/blog/resolving-longstanding-issues-with-wsi

TL;DR: There is now feedback from `vkQueuePresentKHR` about when resources associated with the present operation (i.e the wait semaphores and the swapchain) can be destroyed.

**Comment by krOoze** [comment]
@ShabbyX I read the blog, and scanned the extension, but I am still missing any kind of backwards-looking guidance. What exactly is an application supposed to do when it encounters an implementation that doesn't expose those extensions?

**Comment by krOoze** [comment]
BTW2: surface and swapchain are KHR. Why are these EXT? Are these like a soft-provisional extensions, or?

**Comment by krOoze** [comment]
@HansKristian-Work Precisely. Who's responsibility is to figure out _some_ resolution if not Khronos? It is not reasonable to export the problem to users, which have even less ways to know what to do with it and let them fumble in undefined behavior land. You are uniquely situated to get all the driver implementers into one room, and ask them (considering the spec was fatally defective) wtf did they actually put in their drivers and if they all at least hallucinated similar solutions that are well behaved and from that derive what API users should do.

Either the group needs to explicitly say that it is abandoning all drivers without those extensions as defective, and advise to wholesale blacklist them (which is brutal on Android where people are stuck with old drivers). Or the group needs to provide some "magic" sequence of commands that is known to be well behaved via inner knowledge of the drivers.

> and it has worked, more or less.   
> that have worked successfully in practice.

As I often say, one possible subset of undefined behavior is things look like they work. If it was guaranteed that it always didn't work, then that would be partially-defined behavior.

There are like five and a half vendors. Don't you guys like talk and coordinate stuff at all? Why not simply ask everyone what the deuce they were actually doing with the semaphores if the spec was non-instructive on the point.

It is even partially future-looking thing. New drivers **still** need to implement the core somehow, even if they also implement the extension. If we give guidance "what worked", it might not work when old app (without the extensions) runs on new drivers.

> Unfortunately there isn't a great solution, which is what these extensions fix. FWIW, one workaround is to use the fence passed to vkAcquireNextImageKHR to infer when the resources of the previous call to vkQueuePresentKHR can be destroyed.

Can't do that if you received `VK_ERROR_OUT_OF_DATE_KHR`, and even if it wasn't it is not reasonable to spam `vkAcquireNextImageKHR` when there was not `vkRelease` plus there is no spec guarantee that you will ever get all the image indexes to retire them.

> while vkWaitForPresentKHR, IIUC, waits for the image to actually finish presenting (i.e turning into photons and then be replaced by another image)

That's would be sneaky behavior, because previously I was taught the semaphores are atomic. I mean, logically, if the image was already read, then the semaphore must have been finished waited on. So it feels it should be giving same guarantees.

**Comment by cubanismo** [comment]
I think fixing the spec with something that lets apps behave as initially intended was higher priority than wordsmithing the terrible workarounds apps are allowed to use based on current behavior, so that's what we've done thus far with our limited time.

It may make sense to include vkQueuePresentKHR() in the "submission order" definition of the behavior of fences signaled by vkQueueSubmit(), so as to include it automatically in the WaitIdle() functions as well. The new present fence could also be defined to obey similar submission order rules given the operations it is intended to synchronize (all queue-side operations associated with a given present operation) are similar. To the best of my knowledge, that should let everyone's favorite workarounds for this spec issue continue working. However, it does preclude some good presentation optimizations, so it's not quite a no-brainer IMHO. It might be preferable to define things one way for apps that don't use present fences, and another way for apps that do, which would be a messier spec, but maybe a faster implementation for apps using the new path.

To be air-tight, your app would still also need to separately wait for for any semaphores and fences returned by AcquireNextImage() before destroying a swapchain, as discussed in #1059.

Practically, vkWaitForPresentKHR() is indeed sufficient to determine the semaphores associated with presents are unsignalled per your logic @krOoze, but it is not sufficient for destroying the swapchain, as it says nothing about whether queue work associated with the swapchain images used by the present is complete.

**Comment by ShabbyX** [comment]
> There are only like five and a half vendors. Don't you guys, like, talk and coordinate stuff at all? Why not simply ask everyone what the deuce they were actually doing with the semaphores if the spec was non-instructive on the point?

I mean, we pretty much know how the drivers work, it's just that the (earliest) point in time when you can destroy semaphores was just not queryable. As others mentioned, there's nothing spec-wise to do about that (other than an extension).

> Can't do that if you received VK_ERROR_OUT_OF_DATE_KHR, ...

You sure can, by inferring things from the acquire of the new swapchain. If you are genuinely looking for a workaround, this strategy has worked well for ANGLE: https://chromium.googlesource.com/angle/angle/+/main/src/libANGLE/renderer/vulkan/doc/PresentSemaphores.md

It's not terribly hard to implement, and actually fits nice with the extension (so the implementation can use that as fallback).

**Comment by krOoze** [comment]
> It may make sense to include vkQueuePresentKHR() in the "submission order" definition of the behavior of fences signaled by vkQueueSubmit(), so as to include it automatically in the WaitIdle() functions as well.

Yes please. And while at it, #1059 too.

> The new present fence could also be defined to obey similar submission order rules given the operations it is intended to synchronize

That would be helpful. Consistency is king. Yea it is not great to have 42 different synchronization primitives that behave slightly differently of each other. Speaking of which, it would be great if it was finally made compatible with timeline semaphores too.

> However, it does preclude some good presentation optimizations, so it's not quite a no-brainer IMHO.

I largely think swapchain re-creation and teardown are not really hotspots worthy to be worried about. And as long as the swapchain is alive, people can and probably should recycle the semaphores. Meanwhile `vk*WaitIdle` are convenience commands, not performance. It feels like it should be possible to do without sneaking anything into presentation except minimal book-keeping, and only swallow any possible perf hit at the point the `vk*WaitIdle` is called?

> but it is not sufficient for destroying the swapchain, as it says nothing about whether queue work associated with the swapchain images used by the present is complete.

I didn't even think that was a problem previously, since `vkDestroySwapchainKHR` VUs seemed to be pretty permissive. It only says all uses of acquired images have to be completed, which is easy to ensure.

> I mean, we pretty much know how the drivers work

You insiders might. The rest of us don't. Which is why we need a specification in the first place. If you know how they work, then put it in writing in the specification.

> it's just that the (earliest) point in time when you can destroy semaphores was just not queryable

or the medium time, or the latest time, or any time at all. Basically that means semaphores can NEVER be destroyed or reused NO MATTER what one tries. As soon as you present the semaphore you are screwed, because you are forbidden from even calling `vkDestroyDevice`, without destroying the undestroyable semaphores first.

So there MUST be something spec-wise to do about that that is NOT an extension to fix the spec bug, or the only other formally acceptable alternative would be to declare the whole batch of drivers and apps against that spec as defective and blacklisted, and the old swapchain extension as deprecated so it cant be used.

>  > Can't do that if you received VK_ERROR_OUT_OF_DATE_KHR, ...

> You sure can, by inferring things from the acquire of the new swapchain.

You can infer bobkis after you receive `VK_ERROR_OUT_OF_DATE_KHR`. Once you receive `VK_ERROR_OUT_OF_DATE_KHR`, then it is game over and the swapchain stops talking to you, and you cannot acquire nothing anymore.

**Comment by cubanismo** [comment]
> I didn't even think that was a problem previously, since vkDestroySwapchainKHR VUs seemed to be pretty permissive. It only says all uses of acquired images have to be completed, which is easy to ensure.

Text added as part of maintenance1 clarified this, but the queue operations performed by a QueuePresentKHR() are a use of the acquired images.

**Comment by krOoze** [comment]
> Text added as part of maintenance1 clarified this, but the queue operations performed by a QueuePresentKHR() are a use of the acquired images.

So you broke it even more. Great....
